Encryption.md

```diff
@@ -19,216 +19,544 @@
-a key metadata can be specified, with a maximal length of 512. Key metadata is a free-form
-byte array that can be used by a reader to retrieve the column encryption key. 
-
-Parquet file footer, and its nested structures, contain sensitive information - ranging 
+Sometimes, a hardware acceleration of AES is unavailable (e.g. in Java 8). Then AES crypto 
+operations are implemented in software, and can be somewhat slow, becoming a performance 
+bottleneck in certain workloads. AES CTR is a regular (not authenticated) cipher. It is 
+faster than GCM cipher, since it doesn’t perform integrity verification and doesn’t 
+calculate the authentication tag. Actually, GCM is a combination of CTR cipher and an 
+authentication layer  called GMAC. For applications running without AES acceleration 
+and willing to compromise on content verification, CTR cipher can provide a boost in 
+Parquet write/read throughput. The second Parquet algorithm encrypts the data content 
+(pages) with CTR, and the metadata (Thrift structures) with GCM. This allows to encrypt/decrypt 
+the bulk of the data faster, while still verifying the metadata integrity and making 
+sure the file has not been replaced with a wrong version. However, tampering with the 
+page data might go unnoticed.
+
+The initialization vectors (IVs) in GCM and CTR ciphers must be unique for each encrypted 
+stream. The CTR IVs are comprised of two parts: a unique nonce (“number used once”) and 
+an initial counter field; both must be passed to the CTR cipher upon stream initialization. 
+The GCM mode creates a CTR cipher internally, and passes an IV comprised of a nonce (passed to 
+the GCM cipher upon stream initialization) and an initial counter field constructed internally 
+by the GCM cipher according to the GCM specification (NIST SP 800-38D).
+Parquet encryption uses the RBG-based (random bit generator) nonce construction as defined in 
+the NIST SP 800-38D document, section 8.2.2. Notice: NIST SP 800-38D uses a term “IV” for what 
+is called “nonce” in the Parquet encryption design.
+
+An input to AES CTR encryption is an encryption key, a 16-byte IV and a plaintext. The output
+is a ciphertext with the length equal to that of plaintext (for non-padding encryption).
+
+An input to AES GCM encryption is an encryption key, a 12-byte nonce and a plaintext. The output
+is a ciphertext with the length equal to that of plaintext (for non-padding encryption), and 
+a 16-byte authentication tag used to verify the ciphertext integrity.
+
+
+### 4.1 AES_GCM_V1 encryption algorithm
+This Parquet algorithm encrypts all modules by the GCM cipher, without padding. The AES GCM cipher
+must be implemented by a cryptographic provider according to the NIST SP 800-38D specification. 
+
+Unique nonces are generated for each ciphertext. The nonce length is 12 bytes (96 bits). The NIST 
+specification requires a “random field” to fill all 12 bytes and an arbitrary “free field” 
```

Comment:
What is the "free field"? What is the length of this field? I thought Parquet uses 12 random bytes?

Reply:
"free field" term comes from the NIST spec. The section 8.2.2 says the IVs are comprised of a "random field" and an optional "free field". However, for 96-bit IVs, it requires the random field to be 96-bit long, and "free field" to be empty.

Reply:
The important thing for this section is that the reader understands what the algorithm requires. Then the sections for Parquet modes should specify how to take the 12-byte nonce and pass it to the AES algorithms.